### PR TITLE
Handle rangepos of by-name right-associative applications properly.

### LINKED ISSUE #8562: parameteric applyDynamic creates invalid range positions
```scala
object Test {
  class MyDynamic extends Dynamic {
    def applyDynamic[T](n: String)(a: String) = ???
  }
  val qual = new MyDynamic
  qual.sel[String]("a")
}
```

fails with `-Yrangepos`

```
======= Position error
Enclosing tree [491] does not include tree [490]
== Enclosing tree [491] of type Apply at [135:139]applydynamic_sip.scala

[L   6        ] #491    [135:139]       Apply      // qual.

== Enclosed tree [490] of type TypeApply at [135:150]applydynamic_sip.scala

[L   6        ] #490    [135:150]       TypeApply  // qual.sel[String]
```

Once this is fixed, `-Yrangepos:false` should be removed from `test/files/run/applydynamic_sip.flags`, most likely also `test/files/run/t6731.flags`, `test/files/run/t6663.flags`, `test/files/run/t6327.flags`.

### LINKED ISSUE #10203: RangePos crashes for generic dynamics
I found that I couldn't run scoverage on a project due to compiler crashes. I reduced the issue to the following:

```scala
$ scala -Yrangepos
Welcome to Scala 2.12.1 (OpenJDK 64-Bit Server VM, Java 1.8.0_66).
Type in expressions for evaluation. Or try :help.

scala> import language.dynamics
import language.dynamics

scala> object O extends Dynamic {
     |   def selectDynamic[T](nme: String): String = nme
     | }
defined object O

scala> O.aaaaaaaaaaaaa[Int]
======= Position error
Enclosing tree [6335] does not include tree [6334]
== Enclosing tree [6335] of type Apply at [145:146]<console>

[L  14        ] #6335   [145:146]       Apply      // O.

== Enclosed tree [6334] of type TypeApply at [145:164]<console>

[L  14        ] #6334   [145:164]       TypeApply  // O.aaaaaaaaaaaaa[Int]


While validating #6354
[L   1        ] #6354   [0:174]         PackageDef // $line8

Children:
  [L   1 P# 6354] #6178   [8:14]          Ident      // $line8
  [L   5 P# 6354] #6353   [18:174]        ModuleDef  // $read {
=======
scala.reflect.internal.Positions$ValidateException: Enclosing tree [6335] does not include tree [6334]
	at scala.reflect.internal.Positions.validate$1(Positions.scala:102)
	at scala.reflect.internal.Positions.$anonfun$validatePositions$12(Positions.scala:142)
	at scala.reflect.internal.Positions.validate$1(Positions.scala:142)
	at scala.reflect.internal.Positions.$anonfun$validatePositions$12(Positions.scala:142)
	at scala.reflect.internal.Positions.validate$1(Positions.scala:142)
	at scala.reflect.internal.Positions.$anonfun$validatePositions$12(Positions.scala:142)
	at scala.reflect.internal.Positions.validate$1(Positions.scala:142)
	at scala.reflect.internal.Positions.$anonfun$validatePositions$12(Positions.scala:142)
	at scala.reflect.internal.Positions.validate$1(Positions.scala:142)
	at scala.reflect.internal.Positions.$anonfun$validatePositions$12(Positions.scala:142)
	at scala.reflect.internal.Positions.validate$1(Positions.scala:142)
	at scala.reflect.internal.Positions.$anonfun$validatePositions$12(Positions.scala:142)
	at scala.reflect.internal.Positions.validate$1(Positions.scala:142)
	at scala.reflect.internal.Positions.$anonfun$validatePositions$12(Positions.scala:142)
	at scala.reflect.internal.Positions.validate$1(Positions.scala:142)
	at scala.reflect.internal.Positions.$anonfun$validatePositions$12(Positions.scala:142)
	at scala.reflect.internal.Positions.validate$1(Positions.scala:142)
	at scala.reflect.internal.Positions.$anonfun$validatePositions$12(Positions.scala:142)
	at scala.reflect.internal.Positions.validate$1(Positions.scala:142)
	at scala.reflect.internal.Positions.validatePositions(Positions.scala:147)
	at scala.reflect.internal.Positions.validatePositions$(Positions.scala:83)
	at scala.reflect.internal.SymbolTable.validatePositions(SymbolTable.scala:16)
	at scala.tools.nsc.typechecker.Analyzer$typerFactory$$anon$3.apply(Analyzer.scala:103)
	at scala.tools.nsc.Global$GlobalPhase.$anonfun$applyPhase$1(Global.scala:416)
	at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:409)
	at scala.tools.nsc.typechecker.Analyzer$typerFactory$$anon$3.$anonfun$run$1(Analyzer.scala:94)
	at scala.tools.nsc.typechecker.Analyzer$typerFactory$$anon$3.$anonfun$run$1$adapted(Analyzer.scala:93)
	at scala.collection.Iterator.foreach(Iterator.scala:929)
	at scala.collection.Iterator.foreach$(Iterator.scala:929)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1406)
	at scala.tools.nsc.typechecker.Analyzer$typerFactory$$anon$3.run(Analyzer.scala:93)
	at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1418)
	at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1403)
	at scala.tools.nsc.Global$Run.compileSources(Global.scala:1398)
	at scala.tools.nsc.interpreter.IMain.compileSourcesKeepingRun(IMain.scala:430)
	at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.compileAndSaveRun(IMain.scala:801)
	at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.compile(IMain.scala:759)
	at scala.tools.nsc.interpreter.IMain$Request.compile$lzycompute(IMain.scala:948)
	at scala.tools.nsc.interpreter.IMain$Request.compile(IMain.scala:943)
	at scala.tools.nsc.interpreter.IMain.compile(IMain.scala:599)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:588)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:560)
	at scala.tools.nsc.interpreter.ILoop.interpretStartingWith(ILoop.scala:825)
	at scala.tools.nsc.interpreter.ILoop.command(ILoop.scala:684)
	at scala.tools.nsc.interpreter.ILoop.processLine(ILoop.scala:402)
	at scala.tools.nsc.interpreter.ILoop.loop(ILoop.scala:423)
	at scala.tools.nsc.interpreter.ILoop.$anonfun$process$1(ILoop.scala:993)
	at scala.tools.nsc.interpreter.ILoop.process(ILoop.scala:892)
	at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:75)
	at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:88)
	at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:99)
	at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:104)
	at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)

That entry seems to have slain the compiler.  Shall I replay
your session? I can re-run each line except the last one.
[y/n]
Abandoning crashed session.
```

### LINKED ISSUE #10751: Invalid range pos with right-associative by-name parameters
Found by @hrhino while fixing https://github.com/scala/bug/issues/10706 in https://github.com/scala/scala/pull/6322.

Once this gets fixed, `test/files/pos/t4225.scala` could also get a `-Yrangepos` flags file.

```
$> scalac13 -version
Scala compiler version 2.13.0-M3 -- Copyright 2002-2018, LAMP/EPFL and Lightbend, Inc.
$> cat Test.scala
class C {
  def foo_:(i: => Int): Int = i
  def t(c: C) = 1 foo_: c
}
$> scalac13 -Yrangepos Test.scala
======= Position error
Enclosing tree [122] does not include tree [38]
== Enclosing tree [122] of type Apply at [60:67]Test.scala

[L   3        ] #122    [60:67]         Apply      // foo_: c

== Enclosed tree [38] of type Literal at [58:59]Test.scala

[L   3        ] #38     [58:59]         Literal    // 1


While validating #129
[L   1        ] #129    [0:69]          PackageDef // class C {

Children:
  [L   1 P#  129] #59     [0:0]           Ident      // c
  [L   1 P#  129] #128    [0:69]          ClassDef   // C {
=======
error: scala.reflect.internal.Positions$ValidateException: Enclosing tree [122] does not include tree [38]
```

### PR DESCRIPTION
**Pull request by hrhino**
And, as a bonus, those with infix type args.

The by-name case is a simple tweak to the SIP-34 translation: the pos of the newly-typed `Apply` needs to encompass the inlined rhs of the synthetic `ValDef`. `wrappingPos` was made for this.

For right-associative applications with explicit type args, the situation gets pretty hairy in that we have something akin to:
```
argument operator_: [TYPEARG] qualifier
                    ^-Ident-^
         ^-----------Select-----------^
         ^---------TypeApply----------^
^---------------Apply-----------------^
```
that parses to
```
Apply(
  TypeApply(
    Select(qualifier, operator_:),
    Ident(TYPEARG) :: Nil
  ),
  argument :: Nil
)
```

The problem is that the type argument falls within the `Select`, so both the `fun` and `targs` of the `TypeApply` contain that position. `makeTransparent` was made for this. `makeTransparentIf` was made for concision.

Fixes scala/bug#10751.

Also more accurate positions for dynamic invocations. The `Literal` is positioned at the dynamic name and the `Apply`/`TypeApply` reaches over that.

Fixes scala/bug#10203, and likewise closes scala/bug#8562.

Tweaks the advice given on an invalid assignment to work with rangepos.

Position the result of `mkApplies` during update rewrite.

With this commit and scala/scala#6335 (scala/bug#10733), the compiler
bootstraps with rangepos and all the partests pass with it on.

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Positions.scala [hunk 1]
```diff
@@ -74,7 +74,7 @@ trait Positions extends api.Positions { self: SymbolTable =>
       children foreach (ensureNonOverlapping(_, others, focus))
       if (tree.pos.isOpaqueRange) {
         val wpos = wrappingPos(tree.pos, children, focus)
-        tree setPos (if (isOverlapping(wpos)) tree.pos.makeTransparent else wpos)
+        tree setPos tree.pos.makeTransparentIf(isOverlapping(wpos))
```
Comment (hrhino) [inline src/reflect/scala/reflect/internal/Positions.scala]: overeager

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 2]
```diff
@@ -829,11 +829,14 @@ self =>
...
+        val pos = (opPos | t.pos) makeTransparentIf !treeInfo.isLeftAssoc(op)
         val sel = atPos(pos)(Select(stripParens(t), op.encode))
         if (targs.isEmpty) sel
-        else atPos(pos union targs.last.pos withPoint pos.point) {
-          TypeApply(sel, targs)
+        else {
+          /* if it's right-associative, `targs` are between `op` and `t` so make the pos transparent */
+          atPos((pos ^| targs.last.pos) makeTransparentIf !treeInfo.isLeftAssoc(op)) {
+            TypeApply(sel, targs)
+          }
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: For my own reference in review:

```scala
  def |(that: Position): Position                   = this union that
  def ^(point: Int): Position                       = this withPoint point
  def |^(that: Position): Position                  = (this | that) ^ that.point
  def ^|(that: Position): Position                  = (this | that) ^ this.point
```

File: src/reflect/scala/reflect/internal/util/Position.scala [hunk 3]
```diff
@@ -120,6 +120,8 @@ private[util] trait InternalPositionImpl {
   def isOpaqueRange              = isRange && !isTransparent
   def pointOrElse(alt: Int): Int = if (isDefined) point else alt
   def makeTransparent: Position  = if (isOpaqueRange) Position.transparent(source, start, point, end) else this
+  @inline final def makeTransparentIf(cond: Boolean): Position =
+    if (cond && isOpaqueRange) Position.transparent(source, start, point, end) else this
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/util/Position.scala]: No reason to `@inline` AFAICT.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 4]
```diff
@@ -829,11 +829,14 @@ self =>
         s"Incompatible args to makeBinop: !isExpr but targs=$targs")
 
       def mkSelection(t: Tree) = {
-        val pos = opPos union t.pos
+        val pos = (opPos | t.pos) makeTransparentIf !treeInfo.isLeftAssoc(op)
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: perhaps `val rassoc = !treeInfo.isLeftAssoc(op)`, and reuse below.

**Review by retronym — DISMISSED** [review]
Great work, my comments are only nits, address them if you've got time. The substance of this is already mergeable.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala (hunk 4 above)
Comment (hrhino) [reply src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: Done.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala (hunk 2 above)
Comment (hrhino) [reply src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: I don't "have it in for punctuation", so these were pleasantly expressive to me. If you (of all people!) need to have a cross-reference handy, I'd say that's a sign that they may be less transparent than I had hoped. English names have been restored.

File: src/reflect/scala/reflect/internal/util/Position.scala (hunk 3 above)
Comment (hrhino) [reply src/reflect/scala/reflect/internal/util/Position.scala]: No more `@inline`.

**Review by hrhino — COMMENTED** [review]
Nits are good to pick. I've updated the PR accordingly.

Also, after thinking on it for a bit:
- I moved the left-associativity detection logic to `nme.isLeftAssoc`, since it's about names, not trees.
- I switched my test to assert things about the actual assigned positions, like the `dynamic-*` tests which I had to change, to increase confidence. (I made it print after `typer` rather than `parser` to show the post-desugaring-and-inlining result; if you'd rather I can also or exclusively dump after `parser`.) There are enough of these that maybe a `PositionTest` in `partest-extras` may be justifiable.

Holler if you think the commits are too granular and I'll squash them. I tried for a logical separation but I also try for a reasonable number of commits.

**Comment by hrhino** [comment]
okay, I have gotten the 👍 to do as I will with Scala. Apparently the CLA requires not that they sign it but that I "represent that you have received permission to make Contributions on behalf of that employer", which I have (as of yesterday). 

(also, rebased to make sure of no breakage)

**Comment by SethTisue** [comment]
@hrhino welcome back!

**Comment by retronym** [comment]
I've added a merge from 2.13.x to resolve a merge conflict.

**Review by retronym — APPROVED** [review]
Thanks, keep 'em coming 😄

**Comment by hrhino** [comment]
Thanks! 😄

**Pull request merged**